# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [53]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [ ]:
# Write your answers below:

# a) 1318 days
# b) 2017-07-29 00:00:00
# c) 66.57
# d) 2016-11-19 00:00:00
    2016-12-08 00:00:00
    2014-12-31 00:00:00
    2017-03-03 00:00:00
    2016-01-01 00:00:00
    2015-01-04 00:00:00
    2014-01-23 00:00:00
    2014-11-27 00:00:00
    2015-12-04 00:00:00
    2015-02-23 00:00:00

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [360]:
import mysql.connector

In [362]:
#use a text file to initiate password
password = None
with open(r"C:\Users\somoy\Desktop\Olagoke\Data Science\Iron Hack\Secret.txt", 'r') as file:
    for line in file:
        if line.startswith('mysql:'):
            password = line.split(': ')[1].strip()
            break

In [364]:
#open a connection between python and mysql
cnx = mysql.connector.connect(user = "root", password = password, database = "austin_weather", port = 3306)

In [366]:
#check for connection if still active
cnx.is_connected()

True

In [368]:
#setup a carrier or communicator between the sql and python
cursor = cnx.cursor()

#setup a query in SQL
query = ("SELECT * FROM austin_weather;")

#excecute the query
cursor.execute(query)

#retrieve all rows from the executed query
data = cursor.fetchall()

# Retrieve column names
column_names = cursor.column_names

# Create a DataFrame with the data and column names
df = pd.DataFrame(data, columns=column_names)

# Display the DataFrame with correct column names
print(df.head())

        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [376]:
# Close the cursor and connection
cursor.close()
cnx.close()

In [378]:
cnx.is_connected()

False


# Challenge 1 - Dabsescribe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [382]:
#reviewing the variables and data types
print(df.dtypes)


Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
HumidityLowPercent                    object
SeaLevelPressureHighInches            object
SeaLevelPressureAvgInches             object
SeaLevelPressureLowInches             object
VisibilityHighMiles                   object
VisibilityAvgMiles                    object
VisibilityLowMiles                    object
WindHighMPH                           object
WindAvgMPH                            object
WindGustMPH                           object
PrecipitationSumInches                object
Events                                object
dtype: object


In [384]:
#Examine the descriptive statistics of the numeric variables
print(df.describe())

                      Date    TempHighF     TempAvgF     TempLowF
count                 1319  1319.000000  1319.000000  1319.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.766523    14.045904    14.190648


In [386]:
# Look at the first five rows of all variables to evaluate the categorical variables as well
print(df.head(5))

        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


In [390]:
'Date'
'DewPointAvgF'
'DewPointHighF'
'DewPointLowF'
'Events'
'HumidityAvgPercent'
'HumidityHighPercent'
'HumidityLowPercent'
'PrecipitationSumInches'
'SeaLevelPressureAvgInches'
'SeaLevelPressureHighInches'
'SeaLevelPressureLowInches'
'TempAvgF'
'TempHighF'
'TempLowF'
'VisibilityAvgMiles'
'VisibilityHighMiles'
'VisibilityLowMiles'
'WindAvgMPH'
'WindGustMPH'
'WindHighMPH'

'WindHighMPH'

#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [393]:
df['DewPointAvgF'].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [395]:
# Your observation here
#Pandas treated the column as object instead of integers because of the missing values denoted with - in the data set

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [398]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [400]:
#converting all dataset recognized as object into numeric
for col in wrong_type_columns:
    df[col] = pd.to_numeric(df[col], errors = "coerce")

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [403]:
#check the data type of the whole dataset
print(df.dtypes)

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object


In [405]:
df.sample(20)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
708  2015-11-29         47        45        42           44.0          42.0   
701  2015-11-22         59        47        34           33.0          29.0   
389  2015-01-14         40        39        38           38.0          36.0   
1200 2017-04-04         89        75        60           62.0          55.0   
644  2015-09-26         95        82        68           65.0          59.0   
603  2015-08-16         98        87        75           70.0          67.0   
1093 2016-12-18         37        32        26           18.0          12.0   
1150 2017-02-13         78        71        63           64.0          55.0   
773  2016-02-02         72        62        51           36.0          27.0   
570  2015-07-14         97        86        74           72.0          69.0   
1129 2017-01-23         76        61        46           39.0          36.0   
1290 2017-07-03         99        88        76           74.0          71.0   
887  2016-05-26         90        79        68           78.0          73.0   
871  2016-05-10         91        79        66           74.0          70.0   
248  2014-08-26        101        89        77           71.0          68.0   
798  2016-02-27         73        58        42           52.0          45.0   
236  2014-08-14         98        84        70           68.0          61.0   
448  2015-03-14         78        66        53           53.0          49.0   
866  2016-05-05         89        74        59           56.0          53.0   
1270 2017-06-13         98        87        76           73.0          71.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
708           40.0                 96.0                91.0   
701           26.0                 82.0                56.0   
389           34.0                 92.0                86.0   
1200          50.0                 72.0                52.0   
644           51.0                 84.0                54.0   
603           61.0                 82.0                58.0   
1093          10.0                 58.0                48.0   
1150          42.0                 84.0                63.0   
773           23.0                 37.0                28.0   
570           62.0                 87.0                61.0   
1129          34.0                 65.0                44.0   
1290          67.0                 88.0                62.0   
887           67.0                 93.0                76.0   
871           64.0                 93.0                75.0   
248           62.0                 71.0                52.0   
798           36.0                 82.0                62.0   
236           56.0                 73.0                50.0   
448           44.0                 89.0                61.0   
866           50.0                 72.0                51.0   
1270          68.0                 88.0                63.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
708                 86.0  ...                      30.18   
701                 30.0  ...                      30.32   
389                 79.0  ...                      30.33   
1200                32.0  ...                      29.67   
644                 24.0  ...                      29.93   
603                 33.0  ...                      29.93   
1093                38.0  ...                      30.49   
1150                42.0  ...                      30.05   
773                 18.0  ...                      29.78   
570                 34.0  ...                      29.87   
1129                23.0  ...                      29.94   
1290                35.0  ...                      30.00   
887                 59.0  ...                      29.84   
871                 56.0  ...                      29.87   
248                 33.0  ...                      29.93   
798                 42.0  ...                      30.24   
236    

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [409]:
#select only rows with values
missing_values = df[df.isnull().any(axis=1)]
print(missing_values)

           Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4    2013-12-25         58        50        41           44.0          40.0   
6    2013-12-27         60        53        45           41.0          39.0   
7    2013-12-28         62        51        40           43.0          39.0   
42   2014-02-01         76        66        55           62.0          59.0   
51   2014-02-10         60        48        35           49.0          36.0   
...         ...        ...       ...       ...            ...           ...   
1269 2017-06-12         94        85        75           72.0          68.0   
1288 2017-07-01         98        89        79           76.0          72.0   
1304 2017-07-17         98        88        77           75.0          71.0   
1308 2017-07-21        104        91        77           74.0          68.0   
1313 2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAv

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [413]:
#count the number of rows of the main data set
print(len(df))

#count the number of rows of the missing_values data set
print(len(missing_values))

1319
136


#### Calculate the ratio of missing rows to total rows

In [416]:
#ration of missing rows to the total rows
ratio = print(len(missing_values)/len(df))

0.10310841546626232


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [419]:
#number of missing rows in each column
missing_values_per_column = df.isna().sum()
print(missing_values_per_column)

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [422]:
#Number of missing values in PrecipitationSumInches column in ratio to its total number of rows
ratio_precipitationSumInches = df['PrecipitationSumInches'].isna().sum()/len(df['PrecipitationSumInches'])
print(ratio_precipitationSumInches)

0.09401061410159212


Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [425]:
df.drop(columns="PrecipitationSumInches", inplace=True)

In [427]:
print(df.columns)

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events'],
      dtype='object')


#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [430]:
#Interpolate the missing values
df_fixed = df.interpolate()

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [433]:
print(df.isna().sum())
print(df_fixed.isna().sum())

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64
Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLeve

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [437]:
#evaluating the Events column
print(df["Events"].value_counts())


Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64


Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [440]:
# Your answer: the largest number of event that happend in a single day is Rain


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [443]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
df_fixed[event_list]=0

In [445]:
# remove columns added by mistake
df_fixed.drop(columns= ["event", "event_list"], inplace=True)

KeyError: "['event', 'event_list'] not found in axis"

In [449]:
df_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        0            0  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [452]:
#populating the df with 0 and 1 for the events using loop
for event in event_list:
    df_fixed[event] = df_fixed['Events'].str.contains(event, case=False, na=False).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [455]:
# Your code here
df_fixed.sample(20)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
900  2016-06-08         92        81        70           72.0          70.0   
865  2016-05-04         87        71        54           55.0          51.0   
67   2014-02-26         52        44        36           46.0          33.0   
1226 2017-04-30         78        65        52           54.0          44.0   
57   2014-02-16         76        68        59           61.0          57.0   
1000 2016-09-16         95        85        75           74.0          71.0   
1221 2017-04-25         90        76        62           70.0          65.0   
759  2016-01-19         69        56        43           51.0          47.0   
455  2015-03-21         61        58        55           58.0          56.0   
9    2013-12-30         44        40        35           31.0          26.0   
125  2014-04-25         88        76        63           64.0          54.0   
151  2014-05-21         89        80        70           66.0          64.0   
263  2014-09-10         99        88        77           73.0          71.0   
294  2014-10-11         78        69        59           72.0          62.0   
567  2015-07-11         93        83        73           73.0          70.0   
1267 2017-06-10         93        82        70           69.0          65.0   
73   2014-03-04         50        40        30           36.0          31.0   
44   2014-02-03         46        40        34           38.0          33.0   
1280 2017-06-23        104        91        78           75.0          72.0   
6    2013-12-27         60        53        45           41.0          39.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
900           67.0                 93.0                71.0   
865           47.0                 89.0                57.0   
67            21.0                 83.0                63.0   
1226          35.0                 63.0                43.0   
57            53.0                 90.0                74.0   
1000          68.0                 94.0                69.0   
1221          54.0                 90.0                73.0   
759           41.0                100.0                78.0   
455           54.0                100.0                95.0   
9             21.0                 75.0                60.0   
125           45.0                 68.0                51.0   
151           61.0                 84.0                61.0   
263           66.0                 88.0                63.0   
294           55.0                 97.0                85.0   
567           66.0                 94.0                69.0   
1267          59.0                 84.0                62.0   
73            26.0                 92.0                72.0   
44            28.0                 82.0                74.0   
1280          68.0                 91.0                63.0   
6             37.0                 83.0                65.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
900                 49.0  ...                10.0                 7.0   
865                 25.0  ...                10.0                10.0   
67                  43.0  ...                 8.0                 2.0   
1226                22.0  ...                10.0                 7.0   
57                  57.0  ...                 5.0                 0.0   
1000                44.0  ...                10.0                10.0   
1221                55.0  ...                10.0                10.0   
759                 55.0  ...                 7.0                 2.0   
455                 90.0  ...                 7.0                 2.0   
9                   45.0  ...                10.0                10.0   
125                 34.0  ...                10.0                10.0   
151                 38.0  ...                10.0                10.0   
263                 38.0  ...                10.0                10.0   
294  

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [458]:
#Drop the events column
df_fixed.drop("Events", axis=1, inplace=True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [461]:
# convert the date column from string to numeric values
from datetime import datetime
#Convert the 'Date' column to datetime
df_fixed['Date']=pd.to_datetime(df_fixed['Date'], format= "%Y-%m-%d")
df_fixed['Date']=df_fixed['Date'].apply(lambda x: x.toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [466]:
df_fixed.head(100)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0   735223         74        60        45           67.0          49.0   
1   735224         56        48        39           43.0          36.0   
2   735225         58        45        32           31.0          27.0   
3   735226         61        46        31           36.0          28.0   
4   735227         58        50        41           44.0          40.0   
..     ...        ...       ...       ...            ...           ...   
95  735318         66        60        54           55.0          47.0   
96  735319         82        72        61           68.0          63.0   
97  735320         89        76        63           67.0          61.0   
98  735321         77        64        51           48.0          40.0   
99  735322         79        62        45           51.0          43.0   

    DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0           43.0                 93.0                75.0                57.0   
1           28.0                 93.0                68.0                43.0   
2           23.0                 76.0                52.0                27.0   
3           21.0                 89.0                56.0                22.0   
4           36.0                 86.0                71.0                56.0   
..           ...                  ...                 ...                 ...   
95          30.0                 78.0                57.0                36.0   
96          56.0                100.0                78.0                55.0   
97          49.0                 90.0                64.0                37.0   
98          30.0                 68.0                43.0                18.0   
99          38.0                 83.0                56.0                29.0   

    ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0   ...                 10.0                 7.0                 2.0   
1   ...                 10.0                10.0                 5.0   
2   ...                 10.0                10.0                10.0   
3   ...                 10.0                10.0                 7.0   
4   ...                 10.0                10.0                 7.0   
..  ...                  ...                 ...                 ...   
95  ...                 10.0                10.0                 6.0   
96  ...                 10.0                 7.0                 1.0   
97  ...                 10.0                 8.0                 2.0   
98  ...                 10.0                10.0                10.0   
99  ...                 10.0                10.0                10.0   

    WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0          20.0         4.0         31.0     0    0     1             1  
1          16.0         6.0         25.0     0    0     0             0  
2           8.0         3.0         12.0     0    0     0             0  
3          12.0         4.0         20.0     0    0     0             0  
4          10.0         2.0         16.0     0    0     0             0  
..          ...         ...          ...   ...  ...   ...           ...  
95         22.0        10.0         33.0     0    0     1             0  
96         16.0         8.0         26.0     0    0     1             0  
97         18.0         5.0         27.0     0    0     0             1  
98         13.0         5.0         21.0     0    0     0             0  
99         20.0         8.0         30.0     0    0     0             0  

[100 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [477]:
#select X and y
X = df_fixed.drop(columns="TempAvgF", axis=1)
y = df_fixed["TempAvgF"]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [483]:
#import train test split
from sklearn.model_selection import train_test_split


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [485]:
#Perform train test split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=1)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [500]:
df_fixed.shape

(1319, 23)

In [498]:
#The number of rows that account for 80% of our data
df_fixed.iloc[0:1055]

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1050  736273         72        68        63           68.0          66.0   
1051  736274         76        68        59           67.0          60.0   
1052  736275         71        66        61           69.0          65.0   
1053  736276         69        66        62           66.0          64.0   
1054  736277         65        63        60           62.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1050          59.0                100.0                91.0   
1051          57.0                 93.0                85.0   
1052          61.0                100.0                94.0   
1053          62.0                100.0                92.0   
1054          58.0                100.0                94.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1050                81.0  ...                 10.0                 6.0   
1051                76.0  ...                 10.0                10.0   
1052                87.0  ...                 10.0                 5.0   
1053                84.0  ...                 10.0                 8.0   
1054                87.0  ...                 10.0                 6.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1050                 1.0          7.0         2.0         12.0     0    0   
1051                 5.0         12.0         2.0         20.0     0    0   
1052                 1.0          7.0         2.0         12.0     0    0   
1053                 2.0         10.0         4.0         16.0     0    0   
1054                 1.0         10.0         6.0         17.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1050     1             0  
1051     1             0  
1052     1             0  
1053     1             1  
1054     1             0  

[1055 rows x 23 columns]

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [502]:
#Assigning to X train and X test
X_ts_train = df_fixed.iloc[0:1055]
X_ts_test = df_fixed.iloc[1055:1319]

SyntaxError: invalid syntax (1399383949.py, line 1)

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
